longclaw/shipping/models/processors.py

```diff
@@ -0,0 +1,67 @@
+import json
+import hashlib
+
+from django.utils.encoding import force_bytes, force_text
+from django.core.cache import cache
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models, transaction
+from django.dispatch import receiver
+
+from longclaw.basket.models import BasketItem
+from longclaw.basket.signals import basket_modified
+from polymorphic.models import PolymorphicModel
+from wagtail.admin.edit_handlers import FieldPanel
+
+from ..serializers.locations import AddressSerializer
+from ..signals import address_modified
+
+
+class ShippingRateProcessor(PolymorphicModel):
+    countries = models.ManyToManyField('shipping.Country')
+    
+    rates_cache_timeout = 300
+    def get_rates(self, settings=None, basket_id=None, destination=None):
+        kwargs = dict(settings=settings, basket_id=basket_id, destination=destination)
+        key = self.get_rates_cache_key(**kwargs)
+        rates = cache.get(key)
+        if rates is None:
+            with transaction.atomic():
+                rates = self.process_rates(**kwargs)
+            if rates is not None:
+                cache.set(key, rates, self.rates_cache_timeout)
+        return rates
+    
+    def get_rates_cache_key(self, **kwargs):
+        from longclaw.basket.serializers import BasketItemSerializer
```

Comment:
Still need to solve django.core.exceptions.AppRegistryNotReady: Models aren't loaded yet. from this `from longclaw.basket.serializers import BasketItemSerializer` in the `get_rates_cache_key()` method.

```
vagrant@ubuntu-xenial:/vagrant$ python3 manage.py test
Traceback (most recent call last):
  File "manage.py", line 12, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/apps/registry.py", line 114, in populate
    app_config.import_models()
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/apps/config.py", line 211, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 665, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/vagrant/longclaw/configuration/models.py", line 9, in <module>
    from longclaw.shipping.models import Address
  File "/vagrant/longclaw/shipping/models/__init__.py", line 2, in <module>
    from .processors import *
  File "/vagrant/longclaw/shipping/models/processors.py", line 11, in <module>
    from longclaw.basket.serializers import BasketItemSerializer
  File "/vagrant/longclaw/basket/serializers.py", line 3, in <module>
    from longclaw.products.serializers import ProductVariantSerializer
  File "/vagrant/longclaw/products/serializers.py", line 2, in <module>
    from longclaw.utils import ProductVariant, maybe_get_product_model
  File "/vagrant/longclaw/utils.py", line 6, in <module>
    ProductVariant = apps.get_model(*PRODUCT_VARIANT_MODEL.split('.'))
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/apps/registry.py", line 198, in get_model
    self.check_models_ready()
  File "/home/vagrant/.local/lib/python3.5/site-packages/django/apps/registry.py", line 140, in check_models_ready
    raise AppRegistryNotReady("Models aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Models aren't loaded yet.
vagrant@ubuntu-xenial:/vagrant$ 

```

Reply:
Hi @thenewguy 
Is this error the only outstanding issue on this PR from your point of view? I.E you are happy all development is done?

Reply:
Yes. Ready to move forward on this.  I do not think there is a quick fix here because the model imports the serializer which imports the model =(  If DRF allowed the model as a string it would be an easy fix.  But until then, I think the reliable cache key is worth it